src/openloco/windows/construction/StationTab.cpp

```diff
@@ -0,0 +1,438 @@
+#include "../../companymgr.h"
+#include "../../graphics/image_ids.h"
+#include "../../input.h"
+#include "../../localisation/FormatArguments.hpp"
+#include "../../objects/airport_object.h"
+#include "../../objects/cargo_object.h"
+#include "../../objects/dock_object.h"
+#include "../../objects/objectmgr.h"
+#include "../../objects/road_object.h"
+#include "../../objects/road_station_object.h"
+#include "../../objects/track_object.h"
+#include "../../objects/train_station_object.h"
+#include "../../stationmgr.h"
+#include "../../ui/dropdown.h"
+#include "Construction.h"
+
+using namespace openloco::interop;
+using namespace openloco::map;
+using namespace openloco::map::tilemgr;
+
+namespace openloco::ui::windows::construction::station
+{
+    widget_t widgets[] = {
+        commonWidgets(138, 190, string_ids::stringid_2),
+        make_widget({ 3, 45 }, { 132, 12 }, widget_type::wt_18, 1, 0xFFFFFFFF, string_ids::tooltip_select_station_type),
+        make_widget({ 123, 46 }, { 11, 10 }, widget_type::wt_11, 1, string_ids::dropdown, string_ids::tooltip_select_station_type),
+        make_widget({ 35, 60 }, { 68, 68 }, widget_type::wt_3, 1),
+        make_widget({ 112, 104 }, { 24, 24 }, widget_type::wt_9, 1, image_ids::rotate_object, string_ids::rotate_90),
+        widget_end(),
+    };
+
+    window_event_list events;
+
+    // 0x0049E228
+    static void on_mouse_up(window* self, widget_index widgetIndex)
+    {
+        switch (widgetIndex)
+        {
+            case common::widx::close_button:
+                WindowManager::close(self);
+                break;
+
+            case common::widx::tab_construction:
+            case common::widx::tab_overhead:
+            case common::widx::tab_signal:
+            case common::widx::tab_station:
+                common::switchTab(self, widgetIndex);
+                break;
+
+            case widx::rotate:
+                _constructionRotation++;
+                _constructionRotation = _constructionRotation & 3;
+                _stationCost = 0x80000000;
+                self->invalidate();
+                break;
+        }
+    }
+
+    // 0x0049E249
+    static void on_mouse_down(window* self, widget_index widgetIndex)
+    {
+        switch (widgetIndex)
+        {
+            case widx::station_dropdown:
+            {
+                uint8_t stationCount = 0;
+                while (_stationList[stationCount] != 0xFF)
+                    stationCount++;
+
+                auto widget = self->widgets[widx::station];
+                auto xPos = widget.left + self->x;
+                auto yPos = widget.top + self->y;
+                auto width = widget.width() + 2;
+                auto height = widget.height();
+                dropdown::show(xPos, yPos, width, height, self->colours[1], stationCount, (1 << 7));
+
+                if (_byte_1136063 & (1 << 7))
+                {
+                    for (auto stationIndex = 0; stationIndex < stationCount; stationIndex++)
```

Comment:
We could simplify these 4 loops into a templated function:

```
template<typename obj_type>
void AddStationsToDropdown(const uint8_t stationCount)
{
    for (auto stationIndex = 0; stationIndex < stationCount; stationIndex++)
    {
        auto station = _stationList[stationIndex];
        if (station == _lastSelectedStationType)
             dropdown::set_highlighted_item(stationIndex);

        auto obj = objectmgr::get<obj_type>(station);
        dropdown::add(stationIndex, obj->name);
    }
}
```

`AddStationsToDropdown<airport_object>(stationCount);`

Pretty sure that would work.